### ISSUE TITLE
simdjson 0.9.x ondemand parser fails on dynamic document (possible regression) [fixed in 0.9.3, unfixed in main branch]

### ISSUE THREAD

**Issue by jpalus** [simdjson/simdjson#1521]
Standard disclaimer: C++ is not my strongest side so I might as well be doing something wrong. However this code works perfectly fine with simdjson 0.8.2.

simdjson parses highly dynamic recursive JSON structure -- each field might be present or not and how object is processed depends on its contents. Following reproducer demonstrates possible regression introduced in 0.9.x:

Input document:
```json
{
  "type": "root",
  "nodes": [
    {
      "type": "child",
      "nodes": [
      ]
    },
    {
      "type": "child",
      "name": "child-name",
      "nodes": [
      ]
    }
  ]
}
```
Code:
```cpp
void descend(ondemand::object node) {
  string_view node_type;
  auto type_error = node["type"].get(node_type);
  if (!type_error && node_type == "child") {
    string_view name;
    auto name_error = node["name"].get(name);
    if (!name_error) {
      cout << name << endl;
    } 
    return;
  } else {
    for (ondemand::object child_node : node["nodes"]) {
      descend(child_node);
    }
  }
}

int main(void) {
  ondemand::parser parser;
  padded_string from_file = padded_string::load("test.json");
  ondemand::document file_tree = parser.iterate(from_file);
  descend(file_tree);
}
```
Result with simdjson 0.8.2 is expected:
```
child-name
```
While with simdjson 0.9.1 it gives error:
```
terminate called after throwing an instance of 'simdjson::simdjson_error'
  what():  The JSON document has an improper structure: missing or superfluous commas, braces, missing keys, etc.
[1]    46515 abort (core dumped)  ./test
```
Looks like missing name on first child is somehow responsible for the outcome -- if added to structure it all works fine.

**Comment by lemire** [comment]
@jkeiser Do you want to have a look?

**Comment by lemire** [comment]
@jpalus 

The following seems to work...

```C++
#include <iostream>
#include "simdjson.h"

using namespace simdjson;

void descend(ondemand::object node) {
  if(auto type = node["type"]; type.error() == SUCCESS && type == "child") {
      if(node.find_field("name").error() == simdjson::SUCCESS) {
          std::cout << std::string_view(node["name"]) << std::endl;
      }
  } else {
   for (ondemand::object child_node : node["nodes"]) {
      descend(child_node);
    }
  }
}

int main(void) {
  ondemand::parser parser;
  padded_string json = R"({"type":"root","nodes":[{"type":"child","nodes":[]},{"type":"child","name":"child-name","nodes":[]}]})"_padded;
  std::cout << json << std::endl;
  ondemand::document file_tree = parser.iterate(json);
  descend(file_tree);
}
```

**Comment by lemire** [comment]
The following blows up however...

```C++
#include <iostream>
#include "simdjson.h"

using namespace simdjson;

void descend(ondemand::object node) {
  if(auto type = node.find_field_unordered("type"); type.error() == SUCCESS && type == "child") {
      if(node.find_field_unordered("name").error() == simdjson::SUCCESS) {
          std::cout << std::string_view(node["name"]) << std::endl;
      }
  } else {
   for (ondemand::object child_node : node["nodes"]) {
      descend(child_node);
    }
  }
}

int main(void) {
  ondemand::parser parser;
  padded_string json = R"({"type":"root","nodes":[{"type":"child","nodes":[]},{"type":"child","name":"child-name","nodes":[]}]})"_padded;
  std::cout << json << std::endl;
  ondemand::document file_tree = parser.iterate(json);
  descend(file_tree);
}
```

**Comment by lemire** [comment]
Marking as a bug.

**Comment by lemire** [comment]
@jpalus So there is definitively a bug of some kind. I have reduced it somewhat with PR https://github.com/simdjson/simdjson/pull/1524

We shall fix it soon enough. Thanks for the report.

This is the kind of issue I'd like to patch 0.9 if at all possible.

**Comment by lemire** [comment]
I think that the issue has to do with what happens when `find_field_unordered` is called on a key that is missing. I think it leaves the JSON iterator in a bad state.

**Comment by lemire** [comment]
Ok. I think I have a fix. I will patch 0.9.x.

**Comment by lemire** [comment]
@jpalus   Your description looks accurate.

I am very sorry about this, it appears that our tests were insufficient to catch this issue.

I am not certain when the bug was introduced nor how.

**Comment by lemire** [comment]
@jpalus 

Please upgrade to 0.9.2 https://github.com/simdjson/simdjson/releases/tag/v0.9.2

It should not happen again since we now have tests in place.

**Comment by lemire** [comment]
@jkeiser I am leaving this issue open because I am hoping we can get a better fix in place for 1.0. (A cleaner, simpler and more performant fix.)

**Comment by jpalus** [comment]
Thanks @lemire! Confirmed that 0.9.2 fixes the issue for me.
